### DISCUSSION TITLE
Weekly Recap #4

### DISCUSSION THREAD

**Discussion by kaspth** [kaspth/oaken#34] [Weekly Recaps]
Hey everybody, we're almost a month in now!

I had a call with @swanson and @marcelolx last week, and the existing code we had was great at giving the discussion a lot more dexterity. So we could fairly quickly go yay or nay on what we had.

We ended up with some fairly big changes too, some of these are things we've discussed a bit before.

- We're further splitting the 1-1 between a fixture file and a table, and now we're starting a convention of grouping seeds around a top-level domain object like an Account/Team/Organization ala `test/seeds/accounts/business.rb`. This also means our seeds file are more like `config/routes.rb` then, they take up more lines but are grouped in one file — we can always look at better abstractions later. I've already converted this work in #28
- Following on from #28, I've also added a `test/seeds/data` folder for any pure data tables folks expect to have in their db.
- This new convention should also make it way easier to work on file checksum based caching, e.g. if nothing changed we can skip the seed. I've done some sketches on this, but I'm still figuring out a v1.
- We're making ids optional, so `users.update name: "Kasper"` is possible. This needs a little more work because we now need an alternate unique attribute to find by. I'm doing this in #32

I'm also working on getting us to be able to use Oaken via a package provider of some sorts. I've kept deferring it since I knew it would be kinda pointless without making the changes above. Soon though!

# PRs & Discussions

- @tcannonfodder shared ideas in #31 
- I've started setting up testing with multiple database adapters in #29
- I've added some tests for helper methods in #30, I still don't know quite what I want these to be able to do, but I just wanted these tested
- #27 I refactored our fixture converter a bit so we get Rails generators' (through Thor) built-in diffing for generated files

# Next Team Call

I've just put up the meeting details over here https://github.com/kaspth/oaken/discussions/33

----------------

How's it going so far for everyone? Is it still manageable to follow along? Are y'all still learning Ruby tricks from seeing the code now evolving in a slightly different direction?

**Comment by BakiVernes** [comment]
Hi Kasper,

Can you please expand a bit more on this:

> We're making ids optional, so users.update name: "Kasper" is possible. This needs a little more work because we now need an alternate unique attribute to find by. I'm doing this in https://github.com/kaspth/oaken/pull/32

**Reply by kaspth** [reply]
@BakiVernes sure yeah,

So right now we've got `users.update :kasper, name: "Kasper"` where `:kasper` gets hashed and becomes the id of the created record, in the exact same way that fixtures behave:

```yml
kasper:
  name: "Kasper"
```

While fixtures require that everything added gets a label/id, we don't always need that. So I'm trying to add support for just `users.update name: "Kasper"`, but then we need a different column or columns to compare uniqueness against. Otherwise the internal `find_by` call that `update` does can't find anything to update which breaks the conceptual model we're working on.

Let me know if there's something that's still not clear!

**Reply by BakiVernes** [reply]
So in the PR that you mentioned are you gonna tackle the missing find_by column?

**Reply by kaspth** [reply]
@BakiVernes yes, exactly. I'm still trying to figure out how that's gonna work. I'm also going to be testing to see if our default attributes through `with` could work, ala `users.with id: -> { name }`.

**Comment by manuelfcreis** [comment]
An interesting tweet by Nate Berkopec I saw today 😅 

![Screenshot from 2023-09-01 22-27-34](https://github.com/kaspth/oaken/assets/29898180/5d6878d2-a9b2-443d-8b12-a150a37f4885)


https://twitter.com/nateberkopec/status/1570806217546555401
